### PR TITLE
log_dump.c: Fix hang issue in log dump module

### PR DESCRIPTION
**Pull request by kishore-sn**
If log dump is enabled, and we run heapinfo command, then there is a hang issue. This is because, the heapinfo command locks the kernel heap. But, at the same time heapinfo logs are sent to log dump module which tries to allocate buffer. Since, the heap is already locked by the heapinfo command, it results in a lock-up.

To overcome this issue, upon entering logdump api we will check if the kernel heap is locked before performing the log dump operations which require memory to be allocated or freed. The log dump operation will not be performed as long as the heap is locked. So, any logs printed during this time will not get included in the log dump.

Signed-off-by: Kishore S N <kishore.sn@samsung.com>

### CODE REVIEW

File: os/kernel/log_dump/log_dump.c [hunk 1]
```diff
@@ -135,13 +159,16 @@ int log_dump_init(void)
...
 
 	return LOG_DUMP_OK;
-}
 
-/* Structures used to compress last partially filled block */
-unsigned char *last_comp_block;
-size_t last_comp_block_size;
-int last_comp_block_ptr;
-bool compress_last_block;
+exit_with_outbuf:
```
Comment (jeongchanKim) [inline os/kernel/log_dump/log_dump.c:163]: All failure cases print the same message.
How about printing the log from here?
Then we can reduce the code also.

File: os/kernel/log_dump/log_dump.c [hunk 2]
```diff
@@ -118,12 +124,30 @@ int log_dump_init(void)
...
+		lldbg("memory allocation failure\n");
+		goto exit_with_uncomp;
+	}
+
+	last_comp_block = allocate_compress_buffer(4, CONFIG_LOG_DUMP_CHUNK_SIZE);
+	if (last_comp_block == NULL) {
+		lldbg("memory allocation failure\n");
+		goto exit_with_outbuf;
 	}
+
```
Comment (amogh-samsung) [inline os/kernel/log_dump/log_dump.c:149]: Minor, please avoid extra line spaces

File: os/kernel/log_dump/log_dump.c [hunk 3]
```diff
@@ -118,12 +124,30 @@ int log_dump_init(void)
...
+	out_buf = allocate_compress_buffer(0, CONFIG_LOG_DUMP_CHUNK_SIZE);
+	if (out_buf == NULL) {
+		lldbg("memory allocation failure\n");
+		goto exit_with_uncomp;
+	}
+
+	last_comp_block = allocate_compress_buffer(4, CONFIG_LOG_DUMP_CHUNK_SIZE);
+	if (last_comp_block == NULL) {
+		lldbg("memory allocation failure\n");
+		goto exit_with_outbuf;
```
Comment (amogh-samsung) [inline os/kernel/log_dump/log_dump.c:147]: We are making allocations at init time itself to avoid semaphore contentions later?

**Review by swasthik22 — COMMENTED** [review]
Looks good.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**